#159965306 implement CRUD for articles

meshack-mbuvi opened the pull request:
#### What does this PR do?
- Have CRUD on articles endpoint in Authors Haven REST API.
#### Description of Task to be completed?
- Ensure the following endpoints are working:
    - GET /api/articles/
    - GET /api/articles/{id}
    - POST /api/articles/
    - PUT /api/articles/{id}
    - DELETE /api/articles/{id}
#### How should this be manually tested?
- Clone [this](https://github.com/andela/ah-magnificent6.git) repository
- Create and activate a [virtual environment](https://docs.python.org/3/library/venv.html) inside the project base directory.
- Run the command `git checkout ft-crud-on-articles-159965306`
- Install the project dependencies by running `pip install -r requirements.txt`
- Run the server by issuing the commant `python manage.py runserver`
- Using ([Postman](https://www.getpostman.com/) or your favorite REST API client, 
- Create an account by accessing `api/users/signup/` by sending the payload in the format 
`
{
  "firstname": "admin",
  "username": "admin",
  "email": "admin@gmail.com",
  "password": "password",
} 
`
and 
login on `/api/users/login/` using the same format for the payload above
- test the above endpoints by passing data in the format 
`{
    'title':'title of the article',
    'body':'body of the article',
    'description':'Description of the article'
}`

#### Any background context you want to provide?
 In Authors Haven platform, users will have the capability to create articles and publish them for other people to read. By having this ability, users can showcase their creativity and thoughts by creating and publishing articles in our platform.
#### What are the relevant pivotal tracker stories?
[#159965306 User should be able to create articles](https://www.pivotaltracker.com/story/show/159965306)
#### Screenshots 
post an article
![image](https://user-images.githubusercontent.com/9263906/45281772-8530a880-b4e1-11e8-8c74-db4c0a202117.png)
Retrieve all articles
![image](https://user-images.githubusercontent.com/9263906/45281766-806bf480-b4e1-11e8-8c00-f125f6e41a00.png)

Retrieve a single article
![image](https://user-images.githubusercontent.com/9263906/45281807-aa251b80-b4e1-11e8-9b14-b4e5ad99a988.png)

edit an article
![image](https://user-images.githubusercontent.com/9263906/45281730-5e727200-b4e1-11e8-866c-c18ec3f04f99.png)

Delete an article
![image](https://user-images.githubusercontent.com/9263906/45281822-b3ae8380-b4e1-11e8-852c-575ff58cf833.png)

meshack-mbuvi committed:
[test]: add tests and models for creating new articles

- ensure that all edge cases when creating new articles are covered.

[WIP #159965306]

meshack-mbuvi committed:
[test]: add tests and models for creating new articles

- ensure that all edge cases when creating new articles are covered.

[WIP #159965306]

meshack-mbuvi committed:
[feat]: add imagefield to Article model

- Delete model for images
- Add imageField to Article class

[WIP #159965306]

meshack-mbuvi committed:
[feat]: add imagefield to Article model

- Delete model for images
- Add imageField to Article class

[WIP #159965306]

meshack-mbuvi committed:
[feat](Testing): add tests for creating an article

[WIP #159965306]

meshack-mbuvi committed:
[feat](Testing): add tests for creating an article

[WIP #159965306]

meshack-mbuvi committed:
[feat](Testing): add tests for create, retrieve and delete an article

- Add more tests for retrieveing and deleting an article
- Correct status_code check on test for creating an article

[WIP #159965306]

meshack-mbuvi committed:
[feat](Testing): add tests for create, retrieve and delete an article

- Add more tests for retrieveing and deleting an article
- Correct status_code check on test for creating an article

[WIP #159965306]

meshack-mbuvi committed:
run migrations in travis CI

meshack-mbuvi committed:
run migrations in travis CI

meshack-mbuvi committed:
update requirements.txt file

meshack-mbuvi committed:
update requirements.txt file

meshack-mbuvi committed:
change name of test database

meshack-mbuvi committed:
change name of test database

meshack-mbuvi committed:
update command to run coveralls to point to correct test path

meshack-mbuvi committed:
update command to run coveralls to point to correct test path

meshack-mbuvi committed:
[feat](Testing): add migrations for article app

meshack-mbuvi committed:
[feat](Testing): add migrations for article app

meshack-mbuvi committed:
[feat](testing): add tests for CRUD on articles

- Ensure to have tests for creating articles
- Ensure to have tests for updating an article
- Ensure to have tests for deleting an article
- Ensure to have tests for retrieving all articles and also a single article.

[WIP #159965306]

meshack-mbuvi committed:
[feat](testing): add tests for CRUD on articles

- Ensure to have tests for creating articles
- Ensure to have tests for updating an article
- Ensure to have tests for deleting an article
- Ensure to have tests for retrieving all articles and also a single article.

[WIP #159965306]

meshack-mbuvi committed:
[feat](article): implement crud for articles

- ensure authenticated user can create,update,retrieve and delete an article
- ensure tests for CRUD on article exist.

[Finishes #159965306]

meshack-mbuvi committed:
[feat](article): implement crud for articles

- ensure authenticated user can create,update,retrieve and delete an article
- ensure tests for CRUD on article exist.

[Finishes #159965306]

meshack-mbuvi committed:
[test]:fix failing test

meshack-mbuvi committed:
[test]:fix failing test

parseen254 commented on authors/apps/articles/models.py:
```diff
@@ -0,0 +1,51 @@
+        "Returns a string representation of article title."
+        return self.title
+
+    def _generate_unique_slug(self):
```
@meshack-mbuvi You might be reinventing the wheel here. https://books.agiliq.com/projects/django-orm-cookbook/en/latest/slugfield.html check this article out.

parseen254 commented on .travis.yml:
```diff
@@ -8,19 +8,18 @@ install:
   - coveralls
 
-deploy:
+? deploy
```
@meshack-mbuvi what's happening here?

meshack-mbuvi replied:
Whenever I save the file, the editor is turning that statement to look as you have seen above.But am going to have a workaround about it.

parseen254 commented on authors/apps/articles/renderers.py:
```diff
@@ -0,0 +1,23 @@
+
+
+class ArticleJSONRenderer(JSONRenderer):
+    charset = 'utf-8'
```
Also here checkout http://www.django-rest-framework.org/api-guide/renderers/

parseen254 commented on authors/apps/articles/views.py:
```diff
@@ -0,0 +1,132 @@
+    queryset = Article.objects.all()
+    serializer_class = ArticleSerializer
+    permission_classes = (IsAuthenticated,)
+    renderer_classes = (ArticleJSONRenderer,)
```
Check out inbuilt JSON renderer.

parseen254 commented on authors/apps/articles/views.py:
```diff
@@ -0,0 +1,132 @@
+        return Response(serializer.data, status.HTTP_201_CREATED)
+
+
+class ArticleRetrieveUpdateDestroy(generics.RetrieveUpdateDestroyAPIView):
```
Rename this to article details view.

parseen254 commented on requirements.txt:
```diff
@@ -1,10 +1,29 @@
+autopep8==1.4
```
let us keep our requirements clean.

meshack-mbuvi replied:
Noted.

meshack-mbuvi commented:
On it.

On Mon, Sep 10, 2018 at 12:16 PM, david.parseen <notifications@github.com>
wrote:

> *@parseendavid* requested changes on this pull request.
> ------------------------------
>
> In authors/apps/articles/models.py
> <https://github.com/andela/ah-magnificent6/pull/21#discussion_r216245229>:
>
> > +    description = models.TextField()
> +    author = models.ForeignKey(User, on_delete=models.CASCADE)
> +    created_at = models.DateTimeField(auto_now_add=True)
> +    updated_at = models.DateTimeField(auto_now=True)
> +    published_at = models.DateTimeField(auto_now=True)
> +    slug = models.SlugField(unique=True, editable=False, max_length=140)
> +    favorited = models.NullBooleanField(default=False)
> +    favoritesCount = models.IntegerField(default=0)
> +    image = models.ImageField(
> +        upload_to='static/images', default='static/images/no-img.jpg')
> +
> +    def __str__(self):
> +        "Returns a string representation of article title."
> +        return self.title
> +
> +    def _generate_unique_slug(self):
>
> @meshack-mbuvi <https://github.com/meshack-mbuvi> You might be
> reinventing the wheel here. https://books.agiliq.com/
> projects/django-orm-cookbook/en/latest/slugfield.html check this article
> out.
> ------------------------------
>
> In .travis.yml
> <https://github.com/andela/ah-magnificent6/pull/21#discussion_r216246653>:
>
> >    - coverage report -m
>    - coveralls
>  after_success:
>    - coveralls
>
> -deploy:
> +? deploy
>
> @meshack-mbuvi <https://github.com/meshack-mbuvi> what's happening here?
> ------------------------------
>
> In authors/apps/articles/renderers.py
> <https://github.com/andela/ah-magnificent6/pull/21#discussion_r216247475>:
>
> > @@ -0,0 +1,23 @@
> +import json
> +
> +from rest_framework.renderers import JSONRenderer
> +from rest_framework.utils.serializer_helpers import ReturnDict
> +
> +
> +class ArticleJSONRenderer(JSONRenderer):
> +    charset = 'utf-8'
>
> Also here checkout http://www.django-rest-framework.org/api-guide/
> renderers/
> ------------------------------
>
> In authors/apps/articles/views.py
> <https://github.com/andela/ah-magnificent6/pull/21#discussion_r216249867>:
>
> > +from django.views.generic import ListView
> +from rest_framework.views import APIView
> +from .models import Article
> +from .renderers import ArticleJSONRenderer
> +
> +
> +class ArticleAPIView(generics.ListCreateAPIView):
> +    """
> +    user generics.CreateAPIView to expose parameters to the documentation
> +    method: post
> +    Create a new article
> +    """
> +    queryset = Article.objects.all()
> +    serializer_class = ArticleSerializer
> +    permission_classes = (IsAuthenticated,)
> +    renderer_classes = (ArticleJSONRenderer,)
>
> Check out inbuilt JSON renderer.
> ------------------------------
>
> In authors/apps/articles/views.py
> <https://github.com/andela/ah-magnificent6/pull/21#discussion_r216250012>:
>
> > +        # get user data at this point
> +        article = {
> +            'title': request.data.get('title', None),
> +            'body': request.data.get('body', None),
> +            'description': request.data.get('description', None),
> +            'author': request.user.id
> +        }
> +        # pass article data to the serializer class, check whether the data is
> +        # valid and if valid, save it.
> +        serializer = self.serializer_class(data=article)
> +        serializer.is_valid(raise_exception=True)
> +        serializer.save()
> +        return Response(serializer.data, status.HTTP_201_CREATED)
> +
> +
> +class ArticleRetrieveUpdateDestroy(generics.RetrieveUpdateDestroyAPIView):
>
> Rename this to article details view.
> ------------------------------
>
> In requirements.txt
> <https://github.com/andela/ah-magnificent6/pull/21#discussion_r216250900>:
>
> > @@ -1,10 +1,29 @@
> +autopep8==1.4
>
> let us keep our requirements clean.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/andela/ah-magnificent6/pull/21#pullrequestreview-153680225>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI1bIrfK-vR5GybFqOstTYQwz__jnrk1ks5uZi3ugaJpZM4Wg0Ne>
> .
>

meshack-mbuvi committed:
[feat](articles): refactor code to not reinvent the wheel.

- Change how slug is generated for an article title
- add documentation for articles creation, retrieval,update and deletion
- correct requirements.txt and .travis.yml files

[Finishes #159965306]

meshack-mbuvi committed:
[feat](articles): refactor code to not reinvent the wheel.

- Change how slug is generated for an article title
- add documentation for articles creation, retrieval,update and deletion
- correct requirements.txt and .travis.yml files

[Finishes #159965306]

meshack-mbuvi committed:
[test]: update project requirements.txt file

meshack-mbuvi committed:
Merge branch 'ft-crud-on-articles-159965306' of https://github.com/andela/ah-magnificent6 into ft-crud-on-articles-159965306

meshack-mbuvi committed:
[test]: add more tests to improve coverage

meshack-mbuvi committed:
[feat]: format how articles are retrieved.

meshack-mbuvi committed:
[feat]: remove debugging breakpoints in the article views
